### PR TITLE
Nrf52840 spi 3

### PR DESCRIPTION
**Pull request by ncasaril**
The SPI_3_MASTER interface in NRF52840 has some nice features I felt were missing in mynewt. Primarily because of the 16 or 32MHz speed only spi3m can offer. 

Note that spi3m only works as master, and it only seems to work for non-blocking read/write. Suggestions for making the blocking calls work are most welcome. 

I included the workaround for errata 198 ([nRF52840_Rev_1_Errata_v1.0.pdf](https://infocenter.nordicsemi.com/pdf/nRF52840_Rev_1_Errata_v1.0.pdf)) which may not apply to production boards?

### CODE REVIEW

File: hw/mcu/nordic/nrf52xxx/src/nrf52_periph.c [hunk 1]
```diff
@@ -33,7 +33,7 @@
 #include "bus/drivers/i2c_hal.h"
 #endif
 #endif
-#if MYNEWT_VAL(SPI_0_MASTER) || MYNEWT_VAL(SPI_1_MASTER) || MYNEWT_VAL(SPI_2_MASTER)
+#if MYNEWT_VAL(SPI_0_MASTER) || MYNEWT_VAL(SPI_1_MASTER) || MYNEWT_VAL(SPI_2_MASTER) || MYNEWT_VAL(SPI_2_MASTER)
```
Comment (vrahane) [inline hw/mcu/nordic/nrf52xxx/src/nrf52_periph.c]: Shouldn't SPI_2_MASTER be SPI_3_MASTER above ?

Comment (ncasaril) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_periph.c]: Correct, fix pushed. Thanks! Any insight into the failing travis checks?

**Comment by vrahane** [comment]
```
2019/08/15 23:42:36.268 [INFO] Ignoring /home/travis/gopath/src/github.com/apache/mynewt-core/hw/mcu/stm/stm32f7xx/src/ext/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal_msp_template.c because package dictates it.
2019/08/15 23:42:47.100 [INFO] Ignoring /home/travis/gopath/src/github.com/apache/mynewt-core/hw/mcu/stm/stm32f7xx/src/ext/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal_timebase_rtc_alarm_template.c because package dictates it.
2019/08/15 23:42:47.100 [INFO] Ignoring /home/travis/gopath/src/github.com/apache/mynewt-core/hw/mcu/stm/stm32f7xx/src/ext/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal_timebase_rtc_wakeup_template.c because package dictates it.
2019/08/15 23:42:47.100 [INFO] Ignoring /home/travis/gopath/src/github.com/apache/mynewt-core/hw/mcu/stm/stm32f7xx/src/ext/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal_timebase_tim_template.c because package dictates it.
+rc=0
+[[ 0 -ne 0 ]]
+for target in '${TARGETS}'
+echo 'Building target=telee02_loraping'
Building target=telee02_loraping
+newt build -q -l info telee02_loraping
Error: Syscfg restriction violations detected:
    Setting SPI_3_MASTER(0) requires: MCU_TARGET == "nRF52840"

Setting history (newest -> oldest):
    MCU_TARGET: [hw/bsp/telee02:nRF52832, hw/mcu/nordic/nrf52xxx:]
    SPI_3_MASTER: [hw/mcu/nordic/nrf52xxx:0]
```

**Comment by vrahane** [comment]
I think the problem is nrf52xxx get's used for both nrf52832 and nrf52840 and nrf52832 does not have `SPI_3_MASTER`.

File: hw/mcu/nordic/nrf52xxx/src/hal_spi.c [hunk 2]
```diff
@@ -100,6 +100,46 @@ struct nrf52_hal_spi nrf52_hal_spi1;
...
+        occupied_blocks = (1UL << 8);
+    }
+    else
+    {
+        do {
+            occupied_blocks |= block_flag;
+            block_flag <<= 1;
+            block_addr  += 0x2000;
+        } while ((block_addr < buffer_end_addr) && (block_addr < 0x20012000));
+    }
```
Comment (andrzej-kaczmarek) [inline hw/mcu/nordic/nrf52xxx/src/hal_spi.c]: this code should be adjusted to Mynewt coding style

File: hw/bsp/nordic_pca10056/src/arch/cortex_m4/gcc_startup_nrf52840.s [hunk 3]
```diff
@@ -132,6 +132,7 @@ __isr_vector:
     .long   SPIM3_IRQHandler
     .long   0                           /*Reserved */
     .long   PWM3_IRQHandler
+    .long   0                           /*Reserved */
```
Comment (andrzej-kaczmarek) [inline hw/bsp/nordic_pca10056/src/arch/cortex_m4/gcc_startup_nrf52840.s]: nrf52840 nvic supports 48 vectors so there are 2 entries missing here, not one
also SPIM3 should be moved to last vector

**Comment by ncasaril** [comment]
Thanks @andrzej-kaczmarek. 
Let me know if I should rebase and clean the commits up before we proceed to merge.

File: hw/mcu/nordic/nrf52xxx/src/hal_spi.c [hunk 4]
```diff
@@ -110,21 +110,17 @@ anomaly_198_enable(uint8_t const * p_buffer, size_t buf_len)
...
 
-    if (buf_len == 0)
-    {
+    if (buf_len == 0) {
         return;
     }
     uint32_t buffer_end_addr = ((uint32_t)p_buffer) + buf_len;
     uint32_t block_addr      = ((uint32_t)p_buffer) & ~0x1FFF;
     uint32_t block_flag      = (1UL << ((block_addr >> 13) & 0xFFFF));
     uint32_t occupied_blocks = 0;
```
Comment (andrzej-kaczmarek) [inline hw/mcu/nordic/nrf52xxx/src/hal_spi.c]: local variables definitions should be also moved to top of function

**Comment by andrzej-kaczmarek** [comment]
@ncasaril would be great if you can clean up commits, I'll merge it then. also I added one more comment re coding style.

**Comment by ncasaril** [comment]
@andrzej-kaczmarek coding style fixed and commits squashed. Thanks :)

**Pull request merged**